<a href="https://colab.research.google.com/github/nishah8/dataandpython/blob/main/Copy_of_Sorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning
---



In order to effectively analyse a dataset, often we need to prepare it first.
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* sort by a single column e.g.
  * `df.sort_values("Make") `
* sort by multiple columns e.g.
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make
* sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted*

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [2]:
import pandas as pd
df = pd.read_excel( 'https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true')
print (df.head(5))
print()
sorteddf = df.sort_values("Happiness Score")
print (sorteddf)



       Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938           

### Exercise 2 - sort by multiple columns, display the first 5 rows
---

1. sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order
2. display the first 5 rows of sorted data

**Test output**:  
Records 122, 127, 147, 100, 96

In [6]:
import pandas as pd
df = pd.read_excel( 'https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true')
#question says economy then life expectancy but desired output comes out other way around

sortdf = df.sort_values(by = ["Health (Life Expectancy)","Economy (GDP per Capita)"])
print(sortdf.head(5))

                      Country              Region  Happiness Rank  \
122              Sierra Leone  Sub-Saharan Africa             123   
127                  Botswana  Sub-Saharan Africa             128   
147  Central African Republic  Sub-Saharan Africa             148   
100                 Swaziland  Sub-Saharan Africa             101   
96                    Lesotho  Sub-Saharan Africa              97   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
122            4.507         0.07068                   0.33024  0.95571   
127            4.332         0.04934                   0.99355  1.10464   
147            3.678         0.06112                   0.07850  0.00000   
100            4.867         0.08742                   0.71206  1.07284   
96             4.898         0.09438                   0.37545  1.04103   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
122                   0.00000  0.40840                       

### Exercise 3 - sorting in descending order
---

Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
136, 117, 95, 101, 111 Country and Region columns


In [8]:
import pandas as pd
df = pd.read_excel( 'https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true')
freedom_trustsort = df.sort_values(by=['Freedom','Trust (Government Corruption)'], ascending=False)
print(freedom_trustsort[['Country','Region']].tail(5))

                    Country                           Region
136                  Angola               Sub-Saharan Africa
117                   Sudan               Sub-Saharan Africa
95   Bosnia and Herzegovina       Central and Eastern Europe
101                  Greece                   Western Europe
111                    Iraq  Middle East and Northern Africa


# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values
---

1. read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv
2. check if any NA values exist in the dataframe and print the result
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)



In [10]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)
print (df.isnull().values.any())
print()

df.info()
print()


True

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB



### Exercise 5 - remove null values
---

1. remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. remove all NA values across whole dataframe

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [14]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

x = df.dropna(subset = ["life_satisfaction"])
print(x.info())
print()

y = df.dropna()
print(y.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 352 entries, 613 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               352 non-null    object 
 1   area               352 non-null    object 
 2   date               352 non-null    object 
 3   median_salary      347 non-null    float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        352 non-null    object 
 6   recycling_pct      328 non-null    object 
 7   population_size    350 non-null    float64
 8   number_of_jobs     352 non-null    float64
 9   area_size          272 non-null    float64
 10  no_of_houses       272 non-null    float64
 11  borough_flag       352 non-null    int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 35.8+ KB
None

<class 'pandas.core.frame.DataFrame'>
Int64Index: 267 entries, 613 to 1013
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dty

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])`

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01*

In [15]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

areadup = df.drop_duplicates(subset = ['area'])
print(areadup)

         code                      area        date  median_salary  \
0   E09000001            city of london  1999-12-01        33020.0   
1   E09000002      barking and dagenham  1999-12-01        21480.0   
2   E09000003                    barnet  1999-12-01        19568.0   
3   E09000004                    bexley  1999-12-01        18621.0   
4   E09000005                     brent  1999-12-01        18532.0   
5   E09000006                   bromley  1999-12-01        16720.0   
6   E09000007                    camden  1999-12-01        23677.0   
7   E09000008                   croydon  1999-12-01        19563.0   
8   E09000009                    ealing  1999-12-01        20580.0   
9   E09000010                   enfield  1999-12-01        19289.0   
10  E09000011                 greenwich  1999-12-01        21236.0   
11  E09000012                   hackney  1999-12-01        23249.0   
12  E09000013    hammersmith and fulham  1999-12-01        25000.0   
13  E09000014       

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

In [21]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

def normalise_income(df):
  if df['median_salary'] < 27441:
    return 1
  elif df['median_salary'] < 30932:
    return 2
  else:
    return 3

df["median_salary"] = df.apply(normalise_income, axis=1)
print('max value in median salary column:',df['median_salary'].max(),'\nmin value in median salary column:',df['median_salary'].min())

max value in median salary column: 3 
min value in median salary column: 1


### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [24]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)
print(df.describe())

def normalise_jobs(df):
  if df['number_of_jobs'] < 1.570000e+05:
    return 1
  elif df['number_of_jobs'] < 2.217000e+06:
    return 2
  else:
    return 3

df["number_of_jobs"] = df.apply(normalise_jobs, axis=1)
print()
print('max value in number of jobs column:',df['number_of_jobs'].max(),'\nmin value in number of jobs column:',df['number_of_jobs'].min())

       median_salary  life_satisfaction  population_size  number_of_jobs  \
count    1049.000000         352.000000     1.018000e+03    9.310000e+02   
mean    27977.792183           7.485057     6.042576e+06    3.188095e+06   
std      6412.807487           0.198451     1.526810e+07    8.058302e+06   
min     15684.000000           7.000000     6.581000e+03    4.700000e+04   
25%     23857.000000           7.350000     2.243458e+05    9.450000e+04   
50%     27441.000000           7.510000     2.946035e+05    1.570000e+05   
75%     30932.000000           7.640000     4.630098e+06    2.217000e+06   
max     61636.000000           7.960000     6.643555e+07    3.575000e+07   

          area_size  no_of_houses  borough_flag  
count  6.660000e+02  6.660000e+02   1071.000000  
mean   3.724903e+05  8.814682e+05      0.647059  
std    2.157060e+06  3.690376e+06      0.478108  
min    3.150000e+02  5.009000e+03      0.000000  
25%    2.960000e+03  8.763550e+04      0.000000  
50%    4.323000

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)


def normalise_housing(df):
  if df['no_of_houses'] < 1.024020e+05:
    return 1
  elif df['no_of_houses'] < 1.262760e+05:
    return 2
  else:
    return 3

df["housing_volume"] = df.apply(normalise_housing, axis=1)
print()
print('max value in housing volume column:',df['housing_volume'].max(),'\nmin value in housing volume column:',df['housing_volume'].min())


max value in housing volume column: 3 
min value in housing volume column: 1


### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.


In [5]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)


def normalise_areasize(df):
  if df['area_size'] < 3.724903e+05:
    return 0
  else:
    return 1

df["area_size_normalised"] = df.apply(normalise_areasize, axis=1)
print()
print('first instance of value being 1:',df[df['area_size_normalised']==1].first_valid_index(),'\nfirst instance of value being 0:',df[df['area_size_normalised']== 0].first_valid_index())


first instance of value being 1: 0 
first instance of value being 0: 102
